streams/shared/src/main/scala/zio/stream/internal/ChannelExecutor.scala

```diff
@@ -72,37 +72,37 @@ private[zio] class ChannelExecutor[Env, InErr, InElem, InDone, OutErr, OutElem,
   private[this] final def clearInProgressFinalizer(): Unit =
     inProgressFinalizer = null
 
-  def close(ex: Exit[Any, Any])(implicit trace: Trace): ZIO[Env, Nothing, Any] = {
-    def ifNotNull[R](zio: URIO[R, Any]): URIO[R, Any] =
-      if (zio ne null) zio else ZIO.unit
-
-    val runInProgressFinalizers = {
+  def close(ex: Exit[Any, Any])(implicit trace: Trace): ZIO[Env, Nothing, Any] = ZIO.suspendSucceed {
+    val fs = Array.ofDim[URIO[Env, Exit[Nothing, Any]]](4)
+
+    // NOTE: Each finalizer might have altered the state of the vars, so we need to suspend them all
+    fs(0) = ZIO.suspendSucceed {
+      val close0 = closeLastSubstream
+      closeLastSubstream = null
+      if (close0 eq null) Exit.unit
+      else close0
+    }.exit
+
+    fs(1) = ZIO.suspendSucceed {
       val finalizer = inProgressFinalizer
 
-      if (finalizer ne null)
-        finalizer.ensuring(ZIO.succeed(clearInProgressFinalizer()))
-      else null
-    }
+      if (finalizer eq null) Exit.unit
+      else finalizer.ensuring(ZIO.succeed(clearInProgressFinalizer()))
+    }.exit
 
-    val closeSubexecutors =
-      if (activeSubexecutor eq null) null
+    fs(2) = ZIO.suspendSucceed {
+      if (activeSubexecutor eq null) Exit.unit
       else activeSubexecutor.close(ex)
+    }.exit
 
-    val closeSelf: URIO[Env, Any] = {
+    fs(3) = ZIO.suspendSucceed {
       val selfFinalizers = popAllFinalizers(ex)
 
-      if (selfFinalizers ne null)
-        selfFinalizers.ensuring(ZIO.succeed(clearInProgressFinalizer()))
-      else null
-    }
+      if (selfFinalizers eq null) Exit.unit
+      else selfFinalizers.ensuring(ZIO.succeed(clearInProgressFinalizer()))
+    }.exit
 
-    if ((closeSubexecutors eq null) && (runInProgressFinalizers eq null) && (closeSelf eq null)) null
-    else
-      (
-        ifNotNull(closeSubexecutors).exit <*>
-          ifNotNull(runInProgressFinalizers).exit <*>
-          ifNotNull(closeSelf).exit
-      ).map { case (a, b, c) => a *> b *> c }.uninterruptible.flatMap(ZIO.done(_))
+    ZIO.collectAll(fs).flatMap(Exit.collectAllPar(_).get)
```

Comment:
Im not sure it's a good idea to run in parallel here

Reply:
also the `uninterruptible` is gone?

Reply:
also since the array has fixed size just `a *> b *> c *> d` would do I reckon

Reply:
> Im not sure it's a good idea to run in parallel here

It's not, `Exit.collectAllPar` means it'll process successes sequentially but collect the errors as if they're parallel errors. But I think I'll change it to `Exit.collect` anyways since we're not meant to have failures here (only defects are possible).

> also the uninterruptible is gone?

sh*t, seems I accidentally removed it. Nice catch, thanks!

> also since the array has fixed size just a *> b *> c *> d would do I reckon

Using an array would be more performant AFAICT because we avoid the 2x4 `FlatMap`s. Plus it's easier to extend. WDUT, is it clearer to just zip them together?

Reply:
> also the uninterruptible is gone?

I ended up changing the outer `ZIO.suspendSucceed` to `ZIO.uninterruptible` instead

Reply:
oh yeah I need to get used to that zio vs exit stuff. No you're totally right, let's do it!